File: pathSum2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from typing import List, Optional

class Solution:
    def pathSum(self, root: Optional['TreeNode'], targetSum: int) -> List[List[int]]:
        res, path = [], []

        def dfs(node: Optional['TreeNode'], remain: int) -> None:
            if not node:
                return
            path.append(node.val)
            remain -= node.val

            if not node.left and not node.right and remain == 0:
                res.append(path[:])  # record a copy

            dfs(node.left, remain)
            dfs(node.right, remain)

            path.pop()  # backtrack

        dfs(root, targetSum)
        return res

File: pathsum2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) [][]int {
    var res [][]int
    var path []int

    var dfs func(node *TreeNode, remain int)
    dfs = func(node *TreeNode, remain int) {
        if node == nil {
            return
        }
        path = append(path, node.Val)
        remain -= node.Val

        if node.Left == nil && node.Right == nil && remain == 0 {
            cp := make([]int, len(path))
            copy(cp, path)
            res = append(res, cp)
        }

        dfs(node.Left, remain)
        dfs(node.Right, remain)

        path = path[:len(path)-1] // backtrack
    }

    dfs(root, targetSum)
    return res
}
